lib/seattleflu/id3c/cli/command/etl/redcap.py

```diff
@@ -44,3 +69,616 @@ def normalize_net_id(net_id: str=None) -> Optional[str]:
+
+    if not (location and location.identifier):
+        LOG.debug("No census tract found for given location.")
+        return None
+
+    return create_location(
+        f"{system_identifier}/location/tract", location.identifier, location_type
+        )
+
+
+def build_residential_location_resources(db: DatabaseSession, cache: TTLCache, housing_type: str,
+        primary_street_address: str, secondary_street_address: str, city: str, state: str,
+        zipcode: str, system_identifier: str) -> list:
+    """ Creates a list of residential Location resource entries. """
+
+    lodging_options = [
+        'shelter',
+        'afl',
+        'snf',
+        'ltc',
+        'be',
+        'pst',
+        'cf',
+        'none',
+    ]
+
+    if housing_type in lodging_options:
+        housing_type = 'lodging'
+    else:
+        housing_type = 'residence'
+
+    address = {
+        'street': primary_street_address,
+        'secondary': secondary_street_address,
+        'city': city,
+        'state': state,
+        'zipcode': zipcode
+    }
+
+    lat, lng, canonicalized_address = get_response_from_cache_or_geocoding(address, cache)
+    if not canonicalized_address:
+        return []  # TODO
+
+    tract_location = census_tract(db, (lat, lng), housing_type, system_identifier)
+    # TODO what if tract_location is null?
+    tract_full_url = generate_full_url_uuid()
+    tract_entry = create_resource_entry(tract_location, tract_full_url)
+
+    address_hash = generate_hash(canonicalized_address)
+    address_location = create_location(
+        f"{system_identifier}/location/address",
+        address_hash,
+        housing_type,
+        tract_full_url
+    )
+    address_entry = create_resource_entry(address_location, generate_full_url_uuid())
+
+    return [tract_entry, address_entry]
+
+
+def create_site_reference(default_site: str, system_identifier: str,
+    location: str = None, site_map: dict = None) -> Optional[Dict[str,dict]]:
+    """
+    Create a Location reference for site of the sample collection encounter based
+    on how the sample was collected.
+    """
+    class UnknownRedcapRecordLocation(ValueError):
+        """
+        Raised if a provided *location* is not
+        among a set of expected values.
+        """
+        pass
+
+    if location and site_map:
+        if location not in site_map:
+            raise UnknownRedcapRecordLocation(f"Found unknown location type «{location}»")
+        site = site_map[location]
+    else:
+        site = default_site
+
+    return {
+        "location": create_reference(
+            reference_type = "Location",
+            identifier = create_identifier(f"{system_identifier}/site", site)
+        )
+    }
+
+
+def _create_patient(sex: str, preferred_language: str, record: REDCapRecord,
+        system_identifier: str, first_name: str = None, last_name: str = None,
+        birth_date: str = None, zipcode: str = None, unique_identifier: str = None) -> tuple:
```

Comment:
I wonder here about including the record itself as well as all the parsed fields individually (sex, name, birthdate, etc) -it seems perhaps redundant to include both, rather than just parse out the fields from redcaprecord inside of _create_patient. Do we identify the fields differently from redcap record depending on the caller?

Reply:
The notion of "core" REDCap fields is relatively new. HCT uses them, and we got Childcare to use them. I am having callers pass in particular values as parameters instead of having code inside redcap.py assume field names. I wanted to see the "core" fields stick before I felt comfortable assuming them in utility code.
@joverlee521 had a really neat idea about leveraging a mapping configuration outside the code that would map a project's REDCap fields to "core" fields for projects that didn't confirm to core. That would be neat because then we could safely assume field names inside the utility.

Reply:
Oh I like the mapping config idea. Makes sense, that will make it flexible to be used more broadly for now if we want to use this in older projects also.